### DISCUSSION TITLE
Calculating memory overhead for kubernetes pods

### DISCUSSION THREAD

**Discussion by blkperl** [memcached/memcached#1020] [Q&A] [answered]
Is there a formula or ratio that should be used when calculating memory requests and limits for pods running memcache? We tried adding 10% but it was not enough.

**Comment by dormando** [comment]
What exactly happened when it was "not enough"?

Sorry I'm not familiar with what you're referring to.

**Reply by blkperl** [reply]
Hey @dormando, sorry I should have explained more. The process will be killed by the oomkiller if the request/limit is not correct. In this example memcached used more memory than what it was configured to use. The chart doesn't show memcached using more than the red line but the oomkiller log confirms it used more.

<img width="1086" alt="Screen Shot 2023-04-05 at 7 54 57 AM" src="https://user-images.githubusercontent.com/175551/230120071-ba48e7d4-089b-4af3-afe5-bcbf44269bec.png">

**Reply by dormando** [reply, marked as answer]
It's pretty hard to dig into that from this far away, but the total memory a memcached process can use is:

- The item memory limit from the commandline (-m)
- The hash table (`hash_bytes` in stats output). This can be 1.5x temporarily as the hash table grows.
- Plus some read and log buffers (a handful of megabytes; there are stats counters for those too).

So if you have an instance with a lot of items in it and it decides to expand the hash table, you can run out of memory.

`-o hashpower,nohashexpand` commandline arguments can pre-size and lock the table if you are hyper optimizing memory usage. (see `--help`)
